### PR TITLE
[AArch64] Lower mathlib call ldexp into fscale when sve is enabled

### PR DESCRIPTION
**Pull request by huhu233**
The function of 'fscale' is equivalent to mathlib call ldexp, but has better performance. This patch lowers ldexp into fscale when sve is enabled.

### CODE REVIEW

**Comment by nikic** [comment]
Why does this need special handling in CGP instead of being a normal custom lowering for FLDEXP?

**Comment by paulwalker-arm** [comment]
I agree and custom lowering also gives a straight forward way to support vectors types as well.

**Comment by huhu233** [comment]
> I agree and custom lowering also gives a straight forward way to support vectors types as well.

You are right, I'll update the patch as soon, thanks!

**Comment by huhu233** [comment]
> Why does this need special handling in CGP instead of being a normal custom lowering for FLDEXP?

@nikic , thanks for your suggestion, I'll update the patch as soon!

**Comment by huhu233** [comment]
* Rebase and update the patch

**Review by davemgreen — COMMENTED** [review]
Hi. Can you split this into two separated patches - one for AArch64 and another for X86? I think they should be logically separable.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -57389,3 +57396,40 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+    ExpVT = MVT::v2f64;
+    IID = DAG.getConstant(Intrinsic::x86_avx512_mask_scalef_sd, DL, MVT::i64);
+    break;
+  }
+
+  SDValue Zero = DAG.getConstant(0, DL, MVT::i64);
+  Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XScalarTy, Exp);
+  SDValue VX =
+      DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, XVT, DAG.getUNDEF(XVT), X, Zero);
+  SDValue VExp = DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, ExpVT,
```
Comment (arsenm) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:57427]: Is this lowering the scalar case as a vector operation? Should this move to the vector legalizer?

Comment (huhu233) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:57427]: Hi, @arsenm, here are some of may consideration:

* Is this lowering the scalar case as a vector operation?
Actually, `ldexp@PLT` (ISD::`FLDEXP`) is a scalar call, but `llvm.aarch64.sve.fscale.*` or `llvm.x86.mask.scalef.*` only has vector versions. I hope to replace call with more efficient target supported instructions, so there are some `INSERT_VECTOR_ELT` and `EXTRACT_VECTOR_ELT`. 

* "Should this move to the vector legalizer?" 
`FLDEXP` will occur in many scalar cases, so I think it is not necessary to move to the vector legalizer.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -57389,3 +57396,40 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+
+SDValue X86TargetLowering::LowerFLDEXP(SDValue Op, SelectionDAG &DAG) const {
+  SDValue X = Op.getOperand(0);
+  EVT XScalarTy = X.getValueType();
+  SDValue Exp = Op.getOperand(1);
+
+  SDLoc DL(Op);
+  EVT XVT, ExpVT;
+  SDValue IID;
+  switch (Op.getValueSizeInBits()) {
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:57408]: Thanks, I'll fix it.

File: llvm/test/CodeGen/AArch64/ldexp.ll [hunk 3]
```diff
@@ -0,0 +1,55 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc -O3 -mtriple=aarch64 -mattr=+sve < %s -o - | FileCheck %s
+
+define dso_local nofpclass(nan inf) double @testExp(double noundef nofpclass(nan inf) %val, i32 noundef %a) {
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/test/CodeGen/AArch64/ldexp.ll:4]: OK, thanks for your suggestions!

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 4]
```diff
@@ -26360,3 +26367,39 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
...
+  SDLoc DL(Op);
+  EVT XVT, ExpVT;
+  switch (Op.getValueSizeInBits()) {
+  default:
+    return SDValue();
+  case 32:
+    XVT = MVT::nxv4f32;
+    ExpVT = MVT::nxv4i32;
+    break;
+  case 64:
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:26386]: You are right, but it seems `f16` for `ldexp` is not fully supported? I got some compile failures as shown, 
https://godbolt.org/

**Comment by huhu233** [comment]
* Rebase the branch
* Split the original patch into 2 seperate commits
* Clean up related testcases
* Support `f16` scenario.

**Comment by huhu233** [comment]
ping

**Comment by davemgreen** [comment]
Sorry when I said patch I should have said pull request. Can you split the two patches into separate prs, so they can be reviewed separately. I think what you have for AArch64 looks OK to me

**Comment by huhu233** [comment]
> Sorry when I said patch I should have said pull request. Can you split the two patches into separate prs, so they can be reviewed separately. I think what you have for AArch64 looks OK to me

Hi, @davemgreen , thanks for your reply. Do you mean split these two commits into two Pull Request? If so, I'll remove the commit about X86 in this Patch, and create another Pull Request for it.

**Comment by davemgreen** [comment]
Yeah that sounds good to me, thanks.

**Comment by huhu233** [comment]
* rebase
* Split X86 code into another Pull Request.

**Review by davemgreen — COMMENTED** [review]
Thanks. I checked the edge cases and this looks good to me, if you can address the last round of comments.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 5]
```diff
@@ -26414,3 +26422,46 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
...
+  EVT XScalarTy = X.getValueType();
+  SDValue Exp = Op.getOperand(1);
+
+  SDLoc DL(Op);
+  EVT XVT, ExpVT;
+  switch (Op.getSimpleValueType().SimpleTy) {
+  default:
+    return SDValue();
+  case MVT::f16:
+    X = DAG.getNode(ISD::FP_EXTEND, DL, MVT::f32, X);
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done, thank very much!

File: llvm/test/CodeGen/AArch64/ldexp.ll [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+; CHECK-NEXT:    fcvt h0, s0
+; CHECK-NEXT:    ret
+entry:
+  %0 = tail call fast half @llvm.ldexp.f16.i32(half %val, i32 %a)
+  ret half %0
+}
+
+declare half @llvm.ldexp.f16.i32(half, i32) #1
+
+attributes #1 = { mustprogress nofree nosync nounwind willreturn memory(none) }
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/test/CodeGen/AArch64/ldexp.ll]: Done, only keep the necessary attribute.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 7]
```diff
@@ -26414,3 +26422,46 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
   }
   return true;
 }
+
+SDValue AArch64TargetLowering::LowerFLDEXP(SDValue Op,
```
[the remark this answers is not in the document]
Comment (huhu233) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done, thanks!

**Review by davemgreen — APPROVED** [review]
Thanks, this LGTM.

Do you have commit access, or should I hit commit? If so are you happy for this to go in?

**Comment by huhu233** [comment]
> Thanks, this LGTM.
> 
> Do you have commit access, or should I hit commit? If so are you happy for this to go in?

Hi, @davemgreen, I have asked someone to hit commit, thanks for your reviews!

**Pull request merged**